include/swift/AST/Ownership.h

```diff
@@ -31,21 +33,82 @@ namespace swift {
 enum class ReferenceOwnership : uint8_t {
   /// \brief a strong reference (the default semantics)
   Strong,
+  Default = Strong,
 
-  /// \brief a 'weak' reference
-  Weak,
+#define REF_STORAGE(Name, ...) Name,
+#define REF_STORAGE_RANGE(First, Last) Last_Kind = Last
+#include "swift/AST/ReferenceStorage.def"
+};
+
+enum : unsigned { NumReferenceOwnershipBits =
+  countBitsUsed(static_cast<unsigned>(ReferenceOwnership::Last_Kind)) };
+
+static inline llvm::StringRef keywordOf(ReferenceOwnership ownership) {
```

Comment:
I'm not super happy with dumping all of these in the `swift::` namespace. Maybe there should be a sub-namespace?

Reply:
This seems separable and larger than the pull request at hand. As the person to "blame" for many of these `Num*Bits` enums, I'd be happy to discuss in the forums what the "right" tradeoff is and then implement it. Do you want to start a thread?

Reply:
It's not the Num*Bits I care about, it's `keywordOf` and `Optionality` and things like that. I know the type signature shows what they're for, but this is still a header that gets included all over the place, and it's weird to put these generically-named declarations in the common namespace.

Reply:
I couldn't find any convention in Swift for how to workaround the fact that C++ doesn't allow methods on enum types. I'm open to suggestions.

In the case of this patch series and the way it is evolving, I could probably just move this logic to methods on `ReferenceStorageType`.

Reply:
I think we've just begrudgingly accepted the need to have fairly general-looking function names in the Swift namespace.

Reply:
`Optionality` in particular is really bad, compared to `OptionalTypeKind`.

Reply:
I'm not in love with the name either. I just want to abstract away the hardcoded knowledge about the "optionality" of the reference storage types. How about `OwnershipOptionality`? Or `ReferenceStorageOptionality`?

Reply:
Either of those is better, yeah. I think I'd lean towards the latter—this is not something that needs to be particularly concise.

Reply:
After some thought, I'm going with `OwnershipOptionality`. The rational is threefold:
1) The default `Ownership` is not a reference storage type.
2) An enum over just the reference storage types does not exist and I feel that there is not sufficient demand for such an enum to exist.
3) I don't want to create a situation where the callers of `optionalityOf()` need to check for `Strong` first before calling the API.